src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -892,6 +892,114 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
     }
 }
 
+crate enum PreprocessingError<'a> {
+    Anchor(AnchorFailure),
+    Disambiguator(Range<usize>, String),
+    Resolution(ResolutionFailure<'a>, &'a str, Option<Disambiguator>),
+}
+
+impl From<AnchorFailure> for PreprocessingError<'_> {
+    fn from(err: AnchorFailure) -> Self {
+        Self::Anchor(err)
+    }
+}
+
+crate struct PreprocessingInfo<'a> {
+    crate path_str: &'a str,
+    disambiguator: Option<Disambiguator>,
+    extra_fragment: Option<&'a str>,
+    link_text: String,
+}
+
+/// Returns:
+/// - `None` if the link should be ignored.
+/// - `Some(Err)` if the link should emit an error
+/// - `Some(Ok)` if the link is valid
+///
+/// `link_buffer` is needed for lifetime reasons; it will always be overwritten and the contents ignored.
+crate fn preprocess_link<'a>(
+    ori_link: &'a MarkdownLink,
+    link_buffer_1: &'a mut String,
+    link_buffer_2: &'a mut String,
```

Comment:
`link_buffer` is probably overkill, happy to remove it if you think just copying the string is better.

Reply:
I find it a bit hard to follow the code with `link_buffer`, so I think it might be good to remove it for now. We already introduce copying with things like ``ori_link.link.replace("`", "")``, so it probably isn't that significant of a perf saving.

Reply:
Hmm, I'm not quite show how - I tried adding `to_owned()` in a few places but I still got an error:
```

error[E0515]: cannot return value referencing temporary value
   --> src/librustdoc/passes/collect_intra_doc_links.rs:981:24
    |
935 |       let mut parts = ori_link.link.replace("`", "").split('#');
    |                       ------------------------------ temporary value created here
...
981 |                   return Some(Err(PreprocessingError::Resolution(
    |  ________________________^
982 | |                     err_kind,
983 | |                     path_str.into(),
984 | |                     disambiguator,
985 | |                 )));
    | |___________________^ returns a value referencing data owned by the current function
```
The problem is `path_str` is borrowed from `link`, which is a local `String`, and so I can't return it.

Reply:
Oh never mind I figured it out, `into()` didn't do what I thought.